### ISSUE TITLE
coordinate systems

### ISSUE THREAD

**Issue by alisterburt** [teamtomo/torch-tomogram#22]
related to #3 and relevant for future updates discussed with @davidetorre99 and @McHaillet in #18 and #19 

writing some thoughts down here as I've been thinking too much about coordinate systems the past few days

Clear definitions of coordinate systems feels important for moving forward here, I came up with
- tomogram (arbitrary 3D region of sample space, target for tomogram reconstruction and used for visualization)
- sample (canonical 3D sample space, i.e. sample in microscope before any stage rotation etc, origin at center of sample)
- microscope (fixed 3D coordinate system for microscope, origin in center of sample, tilt axis along y)
- detector (rotated xy plane in microscope, principal axes x and y aligned along detector dimensions)

This feels like overkill when things are simple but is useful to have these layers when thinking about the following problems
- I want to reconstruct a tomogram where my sample is levelled (I include an `M_tomo2sample`)
- I want to know the defoci of a particular point in a tomogram (I query z in microscope space for each image)

We would have the following transforms available on the `TiltSeries`
- `sample2tomo` transforms points in sample space to points in tomograms
- `sample2scope` transforms points from sample space to their final locations in the microscope (tilt axis along y)
- `scope2detector` transforms points from microscope (tilt axis along y) to detector (detector h along y)

This will allow for relatively simple generation of arbitrary rigid body transforms between all spaces, and squishy body stuff remains possible

```python
# rigid body tomogram to position on detector?
M_tomo2sample = torch.linalg.inv(sample2tomo)
M_tomo2detector = M_scope2detector @ M_sample2scope @ M_tomo2sample

point_in_tomogram = torch.tensor([z, y, x, 1]).reshape(4, 1)
point_on_detector = M_tomo2detector @ point_in_tomogram

# how about if we wanted to add sample space volume warping to the projection model?
point_in_sample = M_tomo2sample @ point_in_tomogram
point_in_sample_after_volume_warp = apply_sample_space_deformation(point_in_sample, *deformation_model_params)
point_on_detector = M_scope2detector @ M_sample2scope @ point_in_sample_after_volume_warp
```

for SPA/STA down the road you would have new coordinate systems (not directly relevant for tilt series model)
- particle (3D space with origin at center of object to be reconstructed)
- subparticle (arbitrary coordinate system within an existing particle reconstruction)

this allows for (hopefully) easy use of this machinery to e.g. allow us to reconstruct an oriented subparticle from tilt series

```python
subparticle_position_in_particle_coords = torch.tensor([z, y, x, 1]).reshape(4, 1)

M_particle2detector = M_scope2detector @ M_sample2scope @ M_tomo2sample @ M_particle2tomo
M_subparticle2detector = M_particle2detector @ M_subparticle2particle

subparticle_position_on_detector = M_subparticle2detector @ subparticle_position_in_particle_coords
```

**Comment by McHaillet** [comment]
Very clear what you wrote down! This will make it much easier to stay on the same line in the future :)

> * sample (canonical 3D sample space, i.e. sample in microscope before any stage rotation etc, origin at center of sample)
> 
> * microscope (fixed 3D coordinate system for microscope, origin in center of sample, tilt axis along y)

This is the only part that confuses me. What is the difference between these two? From the definitions they seem almost identical. Is it that the difference in these coordinate systems describes the rotation of the sample in the scope?

**Comment by alisterburt** [comment]
@McHaillet thanks for thinking through! I'm also not 100% about sample vs microscope, the things it felt important to separate were
1. space we actually reconstruct
2. canonical sample in scope coordinates
3. the transformed sample in scope coordinates
3. the transformed sample in detector coordinates

Maybe explaining sample as the internal coordinate system of the sample on the microscope stage feels better?

**Comment by davidetorre99** [comment]
A few thoughts on this to clarify whether I'm understanding correctly and we're all on the same page. I love the idea of defining a "hierarchy" of coordinate systems both for clarity (it's really easy to fall into mistakes here) but also to make life much easier afterwards.
So, when I think about coordinate systems, I mainly think about:
**Tomogram**: Space that we reconstruct where our sample lies and we typically use for visualization/picking/segmentation etc.
**Sample**: Here, I'm perhaps wrong, but I actually see this as an optional space where we specify where the sample sits within the tomogram, namely, the borders of the sample. What I'm not sure about here is whether this space should only set a shift or also a rotation: e.g., tomogram reconstructed without a perfect pretilt, xtilt, in-plane rotation. I think this would add another nice hierarchical layer, but we'd need an input that defines the contour and rotation (e.g. it could be connected to slabify or torch-segment-tomogram-boundaries) or a user given ROI.
**Microscope**: This refers to the positioning of the sample within the microscope. I would actually consider the center as the center of the detector, including rotation and translation with respect to the original position of the sample (at 0 tilt). Essentially, microscope coordinate system and sample one would be identical at 0 tilt and then differ by a rotation and translation for each tilt. 
**Detector**: In this coordinate system we're just projecting the microscope coordinates into a 2D plane → if before we had sample coordinate as **x** = [x, y, z, 1], we would have **x'** = [x', y', z', 1] in the microscope for a tilt image and then **x''** = [x', y'] on the detector.
Then, of course, we can travel between coordinate systems as @alisterburt showed in his examples.
Is that the right way of thinking about these spaces, or am I missing or misinterpreting something?

So essentially what we should have is something along these lines (?):

```
@property
def sample2scope(self) -> torch.Tensor:
    """Transform from sample's intrinsic coords to microscope frame.
    
    Returns:
        (n_tilts, 4, 4) matrices
        
    Physical meaning:
        How the sample (with its internal coordinate system) is 
        positioned and oriented within the fixed microscope frame.
        
    Components:
        1. Ry(tilt_angle): Stage tilt rotation
        2. Rz(tilt_axis_angle): Tilt axis misalignment correction
        3. T(shifts): translations
    """
    R_tilt = Ry(self.tilt_angles, zyx=True, device=self.device)
    R_axis = Rz(self.tilt_axis_angles, zyx=True, device=self.device)
    shifts_3d = F.pad(self.sample_translations / self.pixel_spacing, (1, 0), value=0)
    T_drift = T(shifts_3d, device=self.device)

    # Compose: drift applied after rotations
    return T_drift @ R_axis @ R_tilt

@property
def scope2detector(self) -> torch.Tensor:
    """Transform from microscope frame to detector plane.
    
    Returns: (2, 4) matrix that selects Y and X, drops Z
        
    Physical meaning:
        Pure geometric projection -> detector is fixed perpendicular 
        to beam in microscope frame. This just projects 3D → 2D
        by dropping the Z coordinate (beam direction).
    """
    return torch.tensor([
        [0., 1., 0., 0.],  # Y row
        [0., 0., 1., 0.]   # X row
    ], device=self.device)

@property
def sample2detector(self) -> torch.Tensor:
    """Complete forward projection: sample → scope → detector."""
    # Since scope2detector is just projection (selecting YX rows),
    M_s2sc = self.sample2scope  # (n_tilts, 4, 4)
    # Extract YX rows for projection
    return M_s2sc[:, [1, 2], :]  # (n_tilts, 2, 4)

```

So, for instance if we want to trace a point we can have something like:
```
# Example: Track a point through all spaces

# 1. Point in volume coordinates
point_volume = torch.tensor([100., 50., 0., 1.])  # (z,y,x,1)

# 2. Transform to sample's intrinsic coords
point_sample = self.volume2sample @ point_volume
# "Where is this point in the sample's own coordinate system?"

# 3. Transform to scope (microscope) coords
point_scope = self.sample2scope[tilt_i] @ point_sample  
# "Where is this point in the fixed microscope frame?"

# 4. Project to detector
point_detector = self.scope2detector @ point_scope
# or equivalently:
point_detector = self.sample2detector[tilt_i] @ point_sample
# "Where does this point appear on the 2D detector?"

```

And we can then rewrite project_points as:

```
def project_points(self, points_zyx: torch.Tensor) -> torch.Tensor:
    """Project 3D points from sample coordinates to 2D detector coordinates.
    
    Coordinate flow: sample → scope → detector
    
    Args:
        points_zyx: (..., 3) positions in sample coordinates (Angstroms)
                    Relative to center of sample
    
    Returns:
        (..., n_tilts, 2) positions in detector coordinates (pixels)
                          Relative to center of detector/image
    """
    points_zyx = torch.as_tensor(points_zyx, device=self.device).float()
    
    # Convert from Angstroms to pixels
    points_zyx_px = points_zyx / self.pixel_spacing
    
    # Apply complete sample→detector projection
    # This is: scope2detector @ sample2scope
    M_sample2detector = self.sample2detector  # New property: [:, [1,2], :]
    
    points_zyxw = homogenise_coordinates(points_zyx_px)
    projected_yx = M_sample2detector @ einops.rearrange(
        points_zyxw, "nparticles zyxw -> nparticles 1 zyxw 1"
    )
    projected_yx = einops.rearrange(
        projected_yx, "nparticles ntilts yx 1 -> nparticles ntilts yx"
    )
    return projected_yx
```

Does any of this sort of make sense?

**Comment by alisterburt** [comment]
thanks for writing this up @davidetorre99 

I think you've got it - our mental models are *almost* identical, the only difference I see between our interpretations is that for me 
- the tomogram space should be defined as a subspace of the canonical sample space (you have sample space as relative to tomogram, which makes it feel arbitrary rather than concrete -> to me the tomogram space is arbitrary, the sample space is the fixed global space everything is defined relative to)... 
- when you talk about the microscope coordinate system you immediately refer to transforms, for me:
    - we are defining a computational graph
    - poses are the nodes
    - any transform is an edge in that graph
    - a transform may be within one coordinate system or between coordinate systems

maybe sample space becomes scope0 and scope space becomes scope1 to indicate pre/post microscope rotation? this doesn't mesh nicely with deformations though so I think I still prefer 'sample' and 'scope'

some extra thoughts I wanted to discuss with you:
- we need to clearly define
    1. what is our 'single source of truth' for ts metadata and how is it stored?
    2. where do we keep the image data relative to this metadata
    3. how do we want to handle IO?
    4. how do we handle versioning over time?
- we need to make sure we can add local deformations in a way that is transparent to the user over time


I think this API which clearly separates functionality from metadata could be nice but let me know what you think?
things i like
- stops us from being locked to 4x4 rigid transforms for simple point manipulations
- keeps metadata separate from impl. of functionality

```python
class TiltSeries:
    images: torch.Tensor | None   # (b, h, w)
    metadata: TiltSeriesMetadata  # (paths, alignment metadata)

    @classmethod
    def from_metadata(load_images: bool = False):
        ...

    def extract_particle_tilt_series(positions: torch.Tensor, orientations: torch.Tensor | None) -> tuple[cropped images, orientations]:
        ...

    def tomo2detector(positions: torch.Tensor, orientations: torch.Tensor | None) -> tuple[positions, orientations]:
        ...

    # functions for evaluating other transforms follow same API, including local deformations
    # internally convert to 4x4 where required/useful 

    def load_images():
        ....



```
